README.md

```diff
@@ -104,6 +104,166 @@ TODO: specify the encoding (byte-array to string) procedure
+ * Binary → String:  
+   Generate a canonical textual representation of the given binary input address according to rules of [RFC 5962 section 4](https://tools.ietf.org/html/rfc5952#section-4). Implementations must not produce any of the variations allowed by RFC 4291 mentioned above to ensure that all implementation produce a character by character identical string representation.
+
+Converting between string to binary addresses should be done using the equivalent
+of the POSIX [`inet_pton`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/inet_pton.html)
+and [`inet_ntop`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/inet_ntop.html)
+functions. Alternatively, using the BSD
+[`getaddrinfo`/`freeaddrinfo`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/getaddrinfo.html)
+and [`getnameinfo` with `NI_NUMERICHOST`](https://pubs.opengroup.org/onlinepubs/9699919799/functions/getnameinfo.html)
+may be a viable alternative for some environments.
+
+### `onion`
+
+Encodes a [TOR rendezvous version 2 service pointer](https://gitweb.torproject.org/torspec.git/tree/rend-spec-v2.txt?id=471af27b55ff3894551109b45848f2ce1002441b#n525) (aka .onion-address) and exposed service port on that system.
+
+Protocols using this codec may encode it as binary value of exactly 12 bytes without
+an extra length value.
+
+ * String → Binary:
+    1. Split the input string into 2 parts at the colon character (U+003A COLON):  
+       `(service_str, port_str) = str.split(":")`
+    2. Decode the *service* part before the colon using base32 into binary:  
+       `service_bin = b32decode(service_str)`
+    3. Convert the *port* part to a binary string as specified by the [`uint16be`](#uint16be) codec.
+    4. Concatenate the service and port parts to obtain the final binary encoding:  
+       `return service_bin + port_bin`
+ * Binary → String:
+    1. Split the binary value at the last two bytes into an service name and a port
+       number:  
+       `(service_bin, port_bin) = binary.split_at(-2)`
+    2. Convert the service part into a base32 string:  
+       `service_str = b32encode(service_bin)`
+    3. Convert the *port* part to text as specified by the [`uint16be`](#uint16be) codec.
+    4. Concatenate the result strings using a colon:  
+       `return service_str + ":" + port_str`
+
+### `p2p`
+
+Encodes a libp2p node address.
+
+TBD: Is this really always a base58btc encoded string of at least 5 characters in length!?
```

Comment:
It's anything that parses as a peer ID (the canonical encoding is now actually a libp2p key _CID_): https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md.

Reply:
At least the go implementation does not follow that spec at all but just expects everything to be base58btc: https://github.com/multiformats/go-multiaddr/blob/master/transcoders.go#L293-L315

If your saying that is just a bug to be fixed, I'll be happy to update this text and py-multiaddr accordingly. :slightly_smiling_face:

Reply:
That spec also fails to mention that some IDs are still encoded with `1` as first character, which is also CIDv0

Reply:
@alexander255 correct (well, not a bug so much as something we need to implement).

@ShadowJonathan also correct. I'll take a stab at fixing the spec.

Reply:
@Stebalien: I just tried to implement that spec in Python and there is no guidance in there what to do with binary CIDv0 IDs. I assume they are detected (similar to their string format) using `len(buf) == 34 and buf.startswith(b"\x12\x20")`?